test/src/main/java/org/openjdk/skara/test/HostCredentials.java

```diff
@@ -175,43 +176,78 @@ public HostCredentials(TestInfo testInfo) throws IOException  {
-            var tmpLock = Path.of(credentialsLock + "." + testName + ".tmp");
+    private boolean getLock(HostedRepository repo) throws IOException {
+        try (var tempFolder = new TemporaryDirectory()) {
+            var repoFolder = tempFolder.path().resolve("lock");
+            var lockFile = repoFolder.resolve("lock.txt");
+            Repository localRepo;
             try {
-                Files.writeString(tmpLock, testName);
+                localRepo = Repository.materialize(repoFolder, repo.getUrl(), "testlock");
             } catch (IOException e) {
-                throw new UncheckedIOException(e);
+                // If the branch does not exist, we'll try to create it
+                localRepo = Repository.init(repoFolder, VCS.GIT);
             }
 
-            while (!hasCredentialsLock) {
-                try {
-                    Files.move(tmpLock, credentialsLock);
-                    log.info("Obtained credentials lock for " + testName);
-                    hasCredentialsLock = true;
-                } catch (IOException e) {
-                    log.fine("Failed to obtain credentials lock for " + testName + ", waiting...");
-                    try {
-                        Thread.sleep(Duration.ofSeconds(1).toMillis());
-                    } catch (InterruptedException ignored) {
-                    }
+            if (Files.exists(lockFile)) {
+                var currentLock = Files.readString(lockFile, StandardCharsets.UTF_8);
+                var lockTime = ZonedDateTime.parse(currentLock, DateTimeFormatter.ISO_DATE_TIME);
+                if (lockTime.isBefore(ZonedDateTime.now().minus(Duration.ofMinutes(10)))) {
+                    log.info("Stale lock encountered - overwriting it");
+                } else {
+                    log.info("Active lock encountered - waiting");
+                    return false;
                 }
             }
+
+            // The lock either doesn't exist or is stale, try to grab it
+            Files.writeString(lockFile, ZonedDateTime.now().format(DateTimeFormatter.ISO_DATE_TIME), StandardCharsets.UTF_8);
+            localRepo.add(lockFile);
+            var lockHash = localRepo.commit("Lock", "test", "test@test.test");
+            localRepo.push(lockHash, repo.getUrl(), "testlock");
+            log.info("Obtained credentials lock");
+
+            // If no exception occurs, we have obtained the lock
+            return true;
         }
+    }
+
+    private void releaseLock(HostedRepository repo) throws IOException {
+        try (var tempFolder = new TemporaryDirectory()) {
+            var repoFolder = tempFolder.path().resolve("lock");
+            var lockFile = repoFolder.resolve("lock.txt");
+            Repository localRepo;
+            localRepo = Repository.materialize(repoFolder, repo.getUrl(), "testlock");
+            localRepo.remove(lockFile);
+            var lockHash = localRepo.commit("Lock", "test", "test@test.test");
```

Comment:
Please change the commit message comment to "Unlock" to help future debugging

Reply:
Good point